### PR TITLE
boards: arm: add adc node and fixes for Nucleo L412RB-P

### PR DESCRIPTION
**Pull request by gudnimg**
* Fix wrong SCL pin assignment on Nucleo L412RB-P
* Add SPI support to Nucleo L412RB-P documentation
* SPI pin assignment on Nucleo L412RB-P is now using Arduino pins.
* Add ADC1 node to Nucleo L412RB-P
* ~~Add sensor apds9960 node to two boards (both tested):~~
    * ~~Nucleo L412RB-P~~
    * ~~Nucleo F401RE~~

### CODE REVIEW

**Comment by GunZi200** [comment]
One problem I need to fix... Arduino header uses SPI2 and not SPI1. The L4 device tree does not support SPI2 so I will need to add it myself.

**Comment by GunZi200** [comment]
Added an SPI2 node to L412 device tree and fixed trailing white spaces.

**Comment by GunZi200** [comment]
Added a quick changed: `arduino_spi: &spi1 {};` to `arduino_spi: &spi2 {};`

**Comment by GunZi200** [comment]
Added `#defined(CONFIG_BOARD_NUCLEO_L412RB_P)` to adc test.

**Review by erwango — CHANGES_REQUESTED** [review]
Don't add arbitrary devices to the board configuration.
Board configuration should stick to available "out-of-the-box" hardware.

**Comment by GunZi200** [comment]
> Don't add arbitrary devices to the board configuration.
> Board configuration should stick to available "out-of-the-box" hardware.

Is `.overlay` file in the sensor sample folder a better way? I've not used it before... will look into it 😄

**Comment by GunZi200** [comment]
@erwango please take a look again. I created overlay files on the sample which works.

File: samples/sensor/apds9960/boards/nucleo_f401re.overlay [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2021 Guðni Már Gilbert
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * i2c interface for the Nucleo F401RE and APDS9960
+ */
+
+ &i2c1 {
```
Comment (erwango) [inline samples/sensor/apds9960/boards/nucleo_f401re.overlay:9]: I wonder if there would be ways to set this much more generic using arduino_i2c port reference, and avoid introducing board specific files.
Could you have a check to `samples/sensor/amg88xx/app.overlay` and see if same approach could be used here ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/sensor/apds9960/boards/nucleo_f401re.overlay (hunk 1 above)
Comment (GunZi200) [reply samples/sensor/apds9960/boards/nucleo_f401re.overlay:9]: That would be nice. I will take a look.

Comment (jfischer-no) [reply samples/sensor/apds9960/boards/nucleo_f401re.overlay:9]: app.overlay is not right way here because we have a board in the tree equipped with this sensor. It is better to add a generic shield support for APDS9960 breakout-boards.

**Review by jfischer-no — DISMISSED** [review]

File: samples/sensor/apds9960/boards/nucleo_f401re.overlay (hunk 1 above)
Comment (jfischer-no) [reply samples/sensor/apds9960/boards/nucleo_f401re.overlay:9]: Something like shield support for [SSD1306 displays](https://github.com/zephyrproject-rtos/zephyr/tree/master/boards/shields/ssd1306)

**Comment by erwango** [comment]
> app.overlay is not right way here because we have a board in the tree equipped with this sensor. It is better to add a generic shield support for APDS9960 breakout-boards.

So it means that each time a sensor is available on a supported board, we can't use this generic approach ?
That would be unfortunate.
Is that a pure cmake constraint?

**Comment by GunZi200** [comment]
I'm now facing a problem with the Shield approach.

When I create a shield for the sensor like this (by following https://docs.zephyrproject.org/latest/guides/porting/shields.html): 

image


And then add the shield to the sensor sample's `CMakeLists.txt`:

```cmake
cmake_minimum_required(VERSION 3.13.1)

set(SHIELD adafruit_apds9960) # <-- add SHIELD to preprocessor 

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(apds9960)

FILE(GLOB app_sources src/*.c)
target_sources(app PRIVATE ${app_sources})
```

The sample will work for all boards that support the arduino header. Now the problem I see is how do we determine when to set `SHIELD`?

`set(SHIELD adafruit_apds9960)`

Two cases I see where we don't want this:

1. Board in use already has the sensor onboard (example: Arduino Nano 33 BLE Sense)
2. Board is not compatible with "arduino-header-r3" 

Any ideas on how to check these conditions in CMake? What are your thoughts on this?

**Comment by erwango** [comment]
> Any ideas on how to check these conditions in CMake? What are your thoughts on this?

Please have a look to `samples/shields/*/sample.yaml`:
- provides `    depends_on: arduino_i2c arduino_gpio` filter that will indicate the required arduino connector dependency
- `platform_exclude: disco_l475_iot1` is used to inform that we don't want the sample to be run on some boards (the ones with the same sensor
But I think `platform_exclude` should be replaced by a `filter`:
    filter: !dt_compat_enabled("avago,apds9960")

**Comment by jfischer-no** [comment]
> > app.overlay is not right way here because we have a board in the tree equipped with this sensor. It is better to add a generic shield support for APDS9960 breakout-boards.
> 
> So it means that each time a sensor is available on a supported board, we can't use this generic approach ?
> That would be unfortunate.
> Is that a pure cmake constraint?

Node definition in app.overlay would win and overwrite the one in `boards/arm/reel_board/dts/reel_board.dtsi` for example. But even if it would not, approach with a shield has significantly higher benefit for all.

**Comment by jfischer-no** [comment]
> The sample will work for all boards that support the arduino header. Now the problem I see is how do we determine when to set SHIELD?
> set(SHIELD adafruit_apds9960)
> ```


@GunZi200 
No, the sample should work for all boards equipped with APDS9960 sensor or APDS9960 breakout-board connected.
When APDS9960 breakout-board is connected, the user has to build it with `-DSHIELD=apds9960`. Please do not name it `adafruit_apds9960` or at least do not place it in `boards/shields/adafruit_apds9960`. There are several manufacturers for these breakout-boards, we do not need separate folders for each.

**Comment by erwango** [comment]
> Node definition in app.overlay would win and overwrite the one in boards/arm/reel_board/dts/reel_board.dtsi for example. But even if it would not, approach with a shield has significantly higher benefit for all.

In the long run we need to enable a generic way to run add-on sensors on boards w/o too much code addition (such as having a board.overlay for each board in each sensor with obviously doesn't scale).
Adding a "generic" shield for each sensor/hw piece is somewhat a better proposal, but I consider this is still too much additional code, we need something more efficient.
The `app.overlay` sounds a good approach but I recognize the fact that it does not work with boards that embed the hw makes it broken for now. We should find a way to fix it.

File: boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.dts [hunk 2]
```diff
@@ -53,14 +53,14 @@
 	current-speed = <115200>;
 };
 
-&spi1 {
-	pinctrl-0 = <&spi1_nss_pa4 &spi1_sck_pa5
-			 &spi1_miso_pa6 &spi1_mosi_pa7>;
+&spi2 {
+	pinctrl-0 = <&spi2_nss_pb12 &spi2_sck_pb13
+			 &spi2_miso_pb14 &spi2_mosi_pb15>;
```
Comment (FRASTM) [inline boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.dts:58]: It might be confusing as the nucleo_412rb_p (mb1319) schematic rev C-01 set the PA11 as the SPI_CS and PB12 is not available on the Arduino connector

**Review by FRASTM — APPROVED** [review]

**Comment by jfischer-no** [comment]
@GunZi200 can you please drop the last commit related to APDS9960 sensor so that this PR can be unblocked? You can then open a new PR for APDS9960 shield.

**Comment by GunZi200** [comment]
> @GunZi200 can you please drop the last commit related to APDS9960 sensor so that this PR can be unblocked? You can then open a new PR for APDS9960 shield.

Sure I’ll do it later this evening and also fix the SPI issue @FRASTM found. :)

**Comment by GunZi200** [comment]
I've removed the sensor node changes.

Regarding the SPI CS pin, there is no chip select (NSS) assigned to PA11 in the pinmux or in the datasheet. See:
https://github.com/zephyrproject-rtos/hal_stm32/blob/master/dts/st/l4/stm32l412rbtx-pinctrl.dtsi
```ini
			/* SPI_NSS */

			spi1_nss_pa4: spi1_nss_pa4 {
				pinmux = <STM32_PINMUX('A', 4, AF5)>;
				bias-pull-up;
			};

			spi1_nss_pa15: spi1_nss_pa15 {
				pinmux = <STM32_PINMUX('A', 15, AF5)>;
				bias-pull-up;
			};

			spi1_nss_pb0: spi1_nss_pb0 {
				pinmux = <STM32_PINMUX('B', 0, AF5)>;
				bias-pull-up;
			};

			spi2_nss_pb9: spi2_nss_pb9 {
				pinmux = <STM32_PINMUX('B', 9, AF5)>;
				bias-pull-up;
			};

			spi2_nss_pb12: spi2_nss_pb12 {
				pinmux = <STM32_PINMUX('B', 12, AF5)>;
				bias-pull-up;
			};
```

None of SPI2 NSS are on the arduino connector. I think this is a fault with the board design. But can we not assign any GPIO pin as CS in the device tree? The I/O pin choice should not matter. I'm not sure how to select PA11 specifically.

For now I did not change the pin assignment.

File: boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.dts [hunk 3]
```diff
@@ -78,6 +78,19 @@
...
+&adc1 {
+	/* PA2 and PA3 are used by LPUART1 */
+	pinctrl-0 = <&adc1_in1_pc0 &adc1_in2_pc1
+		&adc1_in3_pc2 &adc1_in4_pc3
+		&adc1_in5_pa0 &adc1_in6_pa1
+		&adc1_in7_pa2 &adc1_in8_pa3
+		&adc1_in9_pa4 &adc1_in10_pa5
+		&adc1_in11_pa6 &adc1_in12_pa7
+		&adc1_in13_pc4 &adc1_in14_pc5
+		&adc1_in15_pb0 &adc1_in16_pb1>;
```
Comment (str4t0m) [inline boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.dts:90]: I'm not so sure about the usefulness of adding so many adc pins.
I think it would be better to just add ~1-3, ideally at header connections usually used for adcs and w/o other conflicts.
In case anyone needs more, they can easily be added using an overlay file.

Comment (GunZi200) [reply boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.dts:90]: > I'm not so sure about the usefulness of adding so many adc pins.
> I think it would be better to just add ~1-3, ideally at header connections usually used for adcs and w/o other conflicts.
> In case anyone needs more, they can easily be added using an overlay file.

That is something I thought about but wasn't sure. I'm guessing potential pin conflicts are the issue or?

Comment (GunZi200) [reply boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.dts]: I removed all but one of the channels and used Arduino pin A0 similar to Nucleo G474RE board.

**Review by MaureenHelm — APPROVED** [review]

File: boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.dts (hunk 2 above)
Comment (erwango) [reply boards/arm/nucleo_l412rb_p/nucleo_l412rb_p.dts:58]: If there's a conflict on revision, please at least add revision information in the board description.
Optionally, you can add multiple revisions support as this is on going [here](https://github.com/zephyrproject-rtos/zephyr/pull/34181)

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by GunZi200** [comment]
@erwango I've added two notes to the documentation. One about the revision number and one about the SPI2 CS pin not being located on the arduino connector.

I'm assuming this is what you meant by:

> If there's a conflict on revision, please at least add revision information in the board description.

**Comment by GunZi200** [comment]
Hi, just a friendly reminder :)

**Comment by MaureenHelm** [comment]
@erwango please revisit

File: dts/arm/st/l4/stm32l412.dtsi [hunk 4]
```diff
@@ -21,6 +21,17 @@
...
+		spi2: spi@40003800 {
+			compatible = "st,stm32-spi";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40003800 0x400>;
+			interrupts = <36 5>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00004000>;
+			status = "disabled";
+			label = "SPI_2";
+		};
```
Comment (erwango) [inline dts/arm/st/l4/stm32l412.dtsi:34]: Please split in a dedicated commit.

File: boards/arm/nucleo_l412rb_p/arduino_r3_connector.dtsi [hunk 5]
```diff
@@ -36,4 +36,4 @@
 };
 
 arduino_i2c: &i2c1 {};
-arduino_spi: &spi1 {};
+arduino_spi: &spi2 {};
```
Comment (erwango) [inline boards/arm/nucleo_l412rb_p/arduino_r3_connector.dtsi:39]: This change is not mentioned in the commit message

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/l4/stm32l412.dtsi (hunk 4 above)
Comment (GunZi200) [reply dts/arm/st/l4/stm32l412.dtsi:34]: Done

File: boards/arm/nucleo_l412rb_p/arduino_r3_connector.dtsi (hunk 5 above)
Comment (GunZi200) [reply boards/arm/nucleo_l412rb_p/arduino_r3_connector.dtsi:39]: Done

**Review by erwango — APPROVED** [review]

**Comment by GunZi200** [comment]
I don’t see how the build failures are related to my changes. Is it possible to re-run it?

**Comment by erwango** [comment]
@GunZi200 Indeed. This should be fixed by rebasing on latest master.
